elastic/beats
# [Heartbeat] Use prctl to ensure subprocesses are cleaned up on exit

Fixes https://github.com/elastic/beats/issues/32363 by instructing the linux kernel to automatically kill node subprocesses if their parents die. In testing it appears chromium always dies as well, although I'm not entirely sure why. Either chrome sets the right flags itself, or the death signal propagates. Either way, in testing this works very solidly.

We don't have sufficient automated test infrastructure to write a good automated test here, so this will have to reply on manual testing.

## Checklist




## How to test this PR locally

Run the below heartbeat inline yaml, this journey will hang for a long time, giving you chrome / node processes that last for a long while. Once it is started, run `pkill heartbeat` which should kill heartbeat and all subprocesses. On the `main` branch heartbeat and chromium will die, and `node` will persiste.

```yaml
- type: browser
  enabled: true
  id: Inline mem 3
  name: Inline mem3
  source:
    inline:
      script:
        step("load homepage", async () => {
            await page.goto('https://www.elastic.co');
        });
        step("hover over products menu", async () => {
            await page.hover('css=[data-nav-item=products]');
        });
        step("failme", async () => {
           await (new Promise(done => {
             setTimeout(done, 100000);
           }));
           await page.hover('css=[data-nav-item=notathingonpage]');
        });
  schedule: "@every 1m"
```
